### ISSUE TITLE
Support of grid-connected Microgrids

### ISSUE THREAD

**Issue by baborok** [Microgrids-X/Microgrids.py#1]
HI do you think i could use your package to simulate such relation in grid ?
Power Grid-Battery-Generator-Load ?
Could not find PowerGrid module in your documentation, maybe i could use PV to simulate it ?

**Comment by pierre-haessig** [comment]
You are right in you diagnosis: there is currently no PowerGrid component  in the package to support grid-connected Microgrids.

The way I see this is:
- If you want to simulate a system with only **Grid-Battery-Renewables**, then you can "repurpose" the [DispatchableGenerator](https://github.com/Microgrids-X/Microgrids.py/blob/main/src/microgrids/components.py#L55)  component to represent the grid. Its "fuel" consumption will happen to be in kWh.
  - for this you need to: set `fuel_intercept` to zero and `fuel_slope` and `fuel_price` to whatever value such that the product  `fuel_slope*fuel_price`  is the price of electricity (and the most readable solution would be to set `fuel_slope=1` so that the fuel volume indeed becomes kWh).
  - however, this only works *if the price of electricity is constant*
  - if the price of electricity is not constant (Time of Use tariffs), this requires creating a new component and also probably adding some code in [`sim_operation`](https://github.com/Microgrids-X/Microgrids.py/blob/main/src/microgrids/operation.py#L160) to record the cumulated cost of grid energy consumption with a time-varying price.
- If you want to simulate a more complex system with **Grid-Battery-Renewables-DispatchableGenerator**, this requires further changes, because this means having to arbitrage between three dispatchable sources (Grid vs DispatchableGenerator vs Battery) while the present version of the [`dispatch`](https://github.com/Microgrids-X/Microgrids.py/blob/main/src/microgrids/operation.py#L92) function only support two (DispatchableGenerator vs Battery).
  - Remark: the same issue will arise to support *several energy storages* and or *several DispatchableGenerators*.

**Comment by baborok** [comment]
Thank you for detailed answer! I will take this recommendations into account.

**Comment by pierre-haessig** [comment]
let's keep the issue open since it's far from being solve. Although it's not in my immediate roadmap, it's definitely a feature worth having.

Out of curiosity, in the list of cases I described, which one are you working on?

**Comment by baborok** [comment]
I have the following problem, there is my conifguration of Grid
Power Grid-Batter-Generator-Load
I need to optimize the capacity of Battery and configuration of Generator(Power output, working hours etc.)
So i will have a certain load and i need to choose an optimal battery and Generator for this load.

**Comment by pierre-haessig** [comment]
and what is the underlying incentive to choose between the Power grid and the Generator:
- is it correct that the Grid is always cheaper, but sometimes it's not available so that the Generator is needed. 
- or is there a time-varying Grid price which makes the Grid sometimes cheaper and sometimes more expensive than the Generator ?

**Comment by baborok** [comment]
Tha main problem is that we are choosing the configuration of generator and batteries so we need to test multiple cases.
Incentive is that generator that we have produces power almost for half a price of main PowerGrid, so our load needs to mainly rely on generator, and batteries are there to optimize this state.
This is rather not a microgrid problem perhaps, but a usual one. We have multiple factories(in our case), that are a clients of company that produces and install mini-gas-generators, and they always struggle to choose the right configuration, because if you pick just oh the top of load level, generator will be heavily underloaded and cost much more, and vice versa if you choose a small amount generator, it will produce less power and you would still need to buy a lot of from main PowerGrid.
I hope you can understand me because english is not my first language, so be free to ask anything you are curios or interested in. :)

**Comment by pierre-haessig** [comment]
OK, thanks for the details. So in your case, an important modeling aspect is the low-load operation of the generator.

For this, Microgrids.py does currently implements an *affine* fuel curve model, like ["F0" in HOMER](https://www.homerenergy.com/products/pro/docs/3.15/generator_fuel_curve_intercept_coefficient.html). This allows having a decreasing generator efficiency at low load when is is >0 (while F0=0 means constant efficiency). The parameter is called in Microgrids.py [`fuel_intercept`](https://github.com/Microgrids-X/Microgrids.py/blob/main/src/microgrids/components.py#L62).

However, Microgrids.py doesn't implement currently the minimal load in the dispatch, although there is already a parameter [`load_ratio_min`](https://github.com/Microgrids-X/Microgrids.py/blob/main/src/microgrids/components.py#L78)  meant for that purpose.

Finally, adding the existence of the external power grid, I guess you need a "smart dispatch" function at each instant to feed the load at the lowest price. As said previously, this is also presently missing.
